### DISCUSSION TITLE
Shelly Plus 0-10V Dimmer

### DISCUSSION THREAD

**Discussion by nullatore** [RavenSystem/esp-homekit-devices#2219] [MEPLHAA Scripts]
Hi,

I'm trying to build up on my own the MEPLHAA script for Shelly Plus 0-10V Dimmer. It'll be little bit over my head but I'm positive. It will a good teaching test.

First of all, here some reference for pinout e for whole characteristic:
- https://templates.blakadder.com/shelly_plus_0-10v_dimmer.html
- https://b2b.shellyitalia.com/p/shelly-plus-0-10v-dimmer/?gad_source=1&gclid=CjwKCAiA_tuuBhAUEiwAvxkgTiqwIOhRgxZGyDGTIfG6ZF7ATR0hbnTxVAP-WWnFTwhSbbhmCw6S5hoCgUgQAvD_BwE

It's a ESP32 cpu base, with a PWM 500Hz Frequency 

(By the way, my goal is to driver a 0-10v led dimmering system who has a 0-10v port)

Here some questions about my very first json (that doesn't work :) )

1) GPIOs Configuration
PWM output (on GPIO26) has to be set as a Hardware o Software PWM?
2) Do I need to configuration PWM Frequency?

```
{
	"c": {"io": [[[0],2],[[26],9],[[25],6,1],[[4,18],6,0,1],[[32],10,0,3]],"y": [500], "l": 0,"b": [[25,5]]},
	"a": [
		{
			"t": 30,
			"b": [[4],[18]],
			"y0":[
					{"v": 0, "0":{ "q": [[26,0,500]]}},
					{"v":20, "0":{ "q": [[26,13107,500]]}},
					{"v":40, "0":{ "q": [[26,26214,500]]}},
					{"v":60, "0":{ "q": [[26,39321,500]]}},
					{"v":80, "0":{ "q": [[26,52428,500]]}},
					{"v":100,"0":{ "q": [[26,65535,500]]}}
				]
		}
	]
}
```

**Comment by nullatore** [comment]
Ok , changing to Software PWM and then set as inverted, it works.

```
{
	"c": {"io": [[[0],2],[[26],7,0,0,1],[[25],6,1],[[4,18],6,0,1],[[32],10,0,3]],"q": 500, "l": 0,"b": [[25,5]]},
	"a": [
		{
			"t": 30,
			"b": [[4],[18]],
			"y0":[
					{"v": 0, "0":{ "q": [[26,0,500]]}},
					{"v":20, "0":{ "q": [[26,13107,500]]}},
					{"v":40, "0":{ "q": [[26,26214,500]]}},
					{"v":60, "0":{ "q": [[26,39321,500]]}},
					{"v":80, "0":{ "q": [[26,52428,500]]}},
					{"v":100,"0":{ "q": [[26,65535,500]]}}
				]
		}
	]
}
```

Next problems :)

I would have a smoother dimmering output. See the red curve.

![image](https://github.com/RavenSystem/esp-homekit-devices/assets/49812905/b7992ead-b03c-4308-b755-4aada5f6334c)

Do I have to increse the items into Y0 section  (v= 0 , v=10, v=20, v=30, ....., v=100)?
Furthermore,
- why Did I have dimmering change at 90% (I declare values 100 and 80...)?
- it seemes that service doesn't consume ON /OFF event but only dimmering events.

Thank you all

**Reply by RavenSystem** [reply]
Don't use wildcards. Use direct PWM control from lightbulb service.

**Comment by RavenSystem** [comment]
Try this:
```json
{
	"c": {"io": [[[0],2],[[26],9,0,65535,1],[[25],6,1],[[4,18],6,0,1],[[32],10,0,3]],"y":[500],"l": 0,"b":[[25,5]]},
	"a": [
		{
			"t": 30,
                        "g":[26],
			"b": [[4],[18]]
		}
	]
}
```

**Reply by nullatore** [reply]
Yeah, it works! Thank you.

Please give me a chance to share another issue.

I would to obtain this beaheavour:
- short press on Buttons GPIO (4,18) : toggles ON / OFF dimmer ---> OK , accomplished 
- long press on Buttons GPIO (4,18) increase o decrease dimmering (swapping themself every long press event)  --> ??

Can you suggest me a way?
